### PR TITLE
[Feature: #162727563] Pagination support for articles

### PR DESCRIPTION
**Pull request by ebenezerdon**
#### What does this PR do?
- This PR adds pagination support for Articles. 

#### Description of Task to be completed?
- Requests can include the number of articles that can be viewed per page and can allow navigation to different pages

#### How should this be manually tested?
- send a `GET` request to `/api/v1/articles/?limit=[amount-of-articles-per-page]&page=[page-number]`

#### What are the relevant pivotal tracker stories?
[#162727563](https://www.pivotaltracker.com/story/show/162727563)

### CODE REVIEW

File: test/004-article.test.js [hunk 1]
```diff
@@ -266,17 +254,36 @@ describe('ARTICLE TEST SUITE', () => {
...
     });
+
+    it('should have amount of returned data as specified', async () => {
+      const response = await chai.request(server)
+        .get('/api/v1/articles/?limit=2');
+      expect(response.status).to.equal(200);
+      expect(response.body.count).to.equal(2);
+    });
+
+    it('should return first page if query.page is less than 1', async () => {
```
Comment (oreoluwade) [inline test/004-article.test.js:275]: You should add specs for two more scenarios
- When there is no `page` in the request.
- When there is no `limit`

The reason for this is that at least one of the branches on line 116 of the article controller is untested.

File: controllers/article.js [hunk 2]
```diff
@@ -4,18 +4,10 @@ import models from '../models';
 import { generateSlug, getReadTime, categories } from '../helpers';
 
 const {
-  Article,
-  User,
-  Profile,
-  Report,
-  Bookmark
+  Article, User, Profile, Report, Bookmark
```
Comment (oreoluwade) [inline controllers/article.js]: If you are going to take them all to the same line, then they should be on the line with `const`. `models` should also be on that line.

Same goes for the destructuring assignment on the next line.

File: test/004-article.test.js (hunk 1 above)
Comment (ebenezerdon) [reply test/004-article.test.js:275]: Okay, I've added the tests

File: controllers/article.js (hunk 2 above)
Comment (ebenezerdon) [reply controllers/article.js]: Eslint does not allow all of them to be on the same line, so I've updated the file to its initial state.

File: test/004-article.test.js [hunk 3]
```diff
@@ -266,17 +254,50 @@ describe('ARTICLE TEST SUITE', () => {
...
     });
 
     it('should not get a non existent article', async () => {
       const response = await chai.request(server)
         .get('/api/v1/articles/non-existing-article');
+      expect(response.status).to.equal(404);
+      expect(response.body.error).to.equal('Article provided does not exist');
+    });
 
+    it('should not get a non existent article', async () => {
```
Comment (oreoluwade) [inline test/004-article.test.js]: This looks to be the same as the spec just before it.

File: controllers/article.js [hunk 4]
```diff
@@ -99,56 +99,46 @@ export default {
...
-            [iLike]: `%${author}%`
-          },
+          '$User.Profile.lastName$': { [iLike]: `%${author}%` },
+          '$User.Profile.firstName$': { [iLike]: `%${author}%` },
+          '$User.username$': { [iLike]: `%${author}%` },
         }
-
       });
     }
+    offset = limit * (page - 1);
```
Comment (oreoluwade) [inline controllers/article.js:130]: What inspired this computation?

Comment (ebenezerdon) [reply controllers/article.js:130]: The `offset` skips a particular number of roles in the database, while the `limit` represents the number of rows to be returned from the database.
So if `page number` is `1`, and `limit` is `5` then the number of roles to be skipped would be `5 * (1-1)` which is `0`. With this, our `db` returns the first 5 rows as page 1.

File: test/004-article.test.js [hunk 5]
```diff
@@ -266,17 +254,50 @@ describe('ARTICLE TEST SUITE', () => {
...
+    });
+
+    it('should return first page if query.page is not provided', async () => {
+      const response = await chai.request(server)
+        .get('/api/v1/articles/?limit=2');
+      expect(response.status).to.equal(200);
+      expect(response.body.page).to.equal(1);
+    });
+
+    it('should return all articles if limit is not provided', async () => {
```
Comment (oreoluwade) [inline test/004-article.test.js:289]: To make this spec an accurate show of the fact that all articles are returned, I think using `page=2`would be great and then expect the count on that page to be 0 or something like that.

**Review by oreoluwade — APPROVED** [review]

**Pull request merged**